README.md
<img align="left" width="440" height="180" alt="Fastest full PostgreSQL nodejs client" src="https://raw.githubusercontent.com/porsager/postgres/master/postgresjs.svg?sanitize=true">

-   [🚀 Fastest full-featured node & deno client](https://github.com/porsager/postgres-benchmarks#results)
-   🏷 ES6 Tagged Template Strings at the core
-   🏄‍♀️ Simple surface API
-   🖊️ Dynamic query support
-   💬 Chat and help on [Gitter](https://gitter.im/porsager/postgres)
-   🐦 Follow on [Twitter](https://twitter.com/rporsager)

<br>

## Getting started

<br>
<img height="220" width="458" alt="Good UX with Postgres.js" src="https://raw.githubusercontent.com/porsager/postgres/master/demo.gif">
<br>

### Installation

```bash
$ npm install postgres
```

### Usage

Create your `sql` database instance

```js
// db.js
import postgres from "postgres";

const sql = postgres({
    /* options */
}); // will use psql environment variables

export default sql;
```

Simply import for use elsewhere

```js
// users.js
import sql from "./db.js";

async function getUsersOver(age) {
    const users = await sql`
    select
      name,
      age
    from users
    where age > ${age}
  `;
    // users = Result [{ name: "Walter", age: 80 }, { name: 'Murray', age: 68 }, ...]
    return users;
}

async function insertUser({ name, age }) {
    const users = await sql`
    insert into users
      (name, age)
    values
      (${name}, ${age})
    returning name, age
  `;
    // users = Result [{ name: "Murray", age: 68 }]
    return users;
}
```

## Table of Contents

-   [Connection](#connection)
-   [Queries](#queries)
-   [Building queries](#building-queries)
-   [Advanced query methods](#advanced-query-methods)
-   [Transactions](#transactions)
-   [Data Transformation](#data-transformation)
-   [Listen & notify](#listen--notify)
-   [Realtime subscribe](#realtime-subscribe)
-   [Numbers, bigint, numeric](#numbers-bigint-numeric)
-   [Result Array](#result-array)
-   [Connection details](#connection-details)
-   [Custom Types](#custom-types)
-   [Teardown / Cleanup](#teardown--cleanup)
-   [Error handling](#error-handling)
-   [TypeScript support](#typescript-support)
-   [Changelog](./CHANGELOG.md)

## Connection

### `postgres([url], [options])`

You can use either a `postgres://` url connection string or the options to define your database connection properties. Options in the object will override any present in the url. Options will fall back to the same environment variables as psql.

```js
const sql = postgres('postgres://username:password@host:port/database', {
  host                 : '',            // Postgres ip address[s] or domain name[s]
  port                 : 5432,          // Postgres server port[s]
  database             : '',            // Name of database to connect to
  username             : '',            // Username of database user
  password             : '',            // Password of database user
  ...and more
})
```

More options can be found in the [Connection details section](#connection-details).

## Queries

### ``await sql`...` -> Result[]``

Postgres.js utilizes [Tagged template functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_templates) to process query parameters **before** interpolation. Using tagged template literals benefits developers by:

1. **Enforcing** safe query generation
2. Giving the ` sql`` ` function powerful [utility](#dynamic-inserts) and [query building](#building-queries) features.

Any generic value will be serialized according to an inferred type, and replaced by a PostgreSQL protocol placeholder `$1, $2, ...`. The parameters are then sent separately to the database which handles escaping & casting.

All queries will return a `Result` array, with objects mapping column names to each row.

```js
const xs = await sql`
  insert into users (
    name, age
  ) values (
    'Murray', 68
  )

  returning *
`;

// xs = [{ user_id: 1, name: 'Murray', age: 68 }]
```

> Please note that queries are first executed when `awaited` – or instantly by using [`.execute()`](#execute).

### Query parameters

Parameters are automatically extracted and handled by the database so that SQL injection isn't possible. No special handling is necessary, simply use tagged template literals as usual.

```js
const name = "Mur",
    age = 60;

const users = await sql`
  select
    name,
    age
  from users
  where
    name like ${name + "%"}
    and age > ${age}
`;
// users = [{ name: 'Murray', age: 68 }]
```

> Be careful with quotation marks here. Because Postgres infers column types, you do not need to wrap your interpolated parameters in quotes like `'${name}'`. This will cause an error because the tagged template replaces `${name}` with `$1` in the query string, leaving Postgres to do the interpolation. If you wrap that in a string, Postgres will see `'$1'` and interpret it as a string as opposed to a parameter.

### Dynamic column selection

```js
const columns = ['name', 'age']

sql`
  select
    ${ sql(columns) }
  from users
`

// Which results in:
select "name", "age" from users
```

### Dynamic inserts

```js
const user = {
  name: 'Murray',
  age: 68
}

sql`
  insert into users ${
    sql(user, 'name', 'age')
  }
`

// Which results in:
insert into users ("name", "age") values ($1, $2)
```

**You can omit column names and simply execute `sql(user)` to get all the fields from the object as columns**. Be careful not to allow users to supply columns that you do not want to be inserted.

#### Multiple inserts in one query

If you need to insert multiple rows at the same time it's also much faster to do it with a single `insert`. Simply pass an array of objects to `sql()`.

```js
const users = [{
  name: 'Murray',
  age: 68,
  garbage: 'ignore'
},
{
  name: 'Walter',
  age: 80
}]

sql`insert into users ${ sql(users, 'name', 'age') }`

// Is translated to:
insert into users ("name", "age") values ($1, $2), ($3, $4)

// Here you can also omit column names which will use object keys as columns
sql`insert into users ${ sql(users) }`

// Which results in:
insert into users ("name", "age") values ($1, $2), ($3, $4)
```

### Dynamic columns in updates

This is also useful for update queries

```js
const user = {
  id: 1,
  name: 'Murray',
  age: 68
}

sql`
  update users set ${
    sql(user, 'name', 'age')
  }
  where user_id = ${ user.id }
`

// Which results in:
update users set "name" = $1, "age" = $2 where user_id = $3
```

### Multiple updates in one query

It's possible to create multiple udpates in a single query. It's necessary to use arrays intead of objects to ensure the order of the items so that these correspond with the column names.

```js
const users = [
    [1, "John", 34],
    [2, "Jane", 27],
];

sql`
  update users set name = update_data.name, (age = update_data.age)::int
  from (values ${sql(users)}) as update_data (id, name, age)
  where users.id = (update_data.id)::int
  returning users.id, users.name, users.age
`;
```

### Dynamic values and `where in`

Value lists can also be created dynamically, making `where in` queries simple too.

```js
const users = await sql`
  select
    *
  from users
  where age in ${sql([68, 75, 23])}
`;
```

or

```js
const [{ a, b, c }] => await sql`
  select
    *
  from (values ${ sql(['a', 'b', 'c']) }) as x(a, b, c)
`
```

## Building queries

Postgres.js features a simple dynamic query builder by conditionally appending/omitting query fragments.
It works by nesting ` sql`` ` fragments within other ` sql`` ` calls or fragments. This allows you to build dynamic queries safely without risking sql injections through usual string concatenation.

### Partial queries

```js
const olderThan = x => sql`and age > ${ x }`

const filterAge = true

sql`
  select
   *
  from users
  where name is not null ${
    filterAge
      ? olderThan(50)
      : sql``
  }
`
// Which results in:
select * from users where name is not null
// Or
select * from users where name is not null and age > 50
```

### Dynamic filters

```js
sql`
  select
    *
  from users ${
    id
      ? sql`where user_id = ${ id }`
      : sql``
  }
`


CHANGELOG.md
# Changelog

## v3.2.4 - 25 May 2022

-   Allow setting keep_alive: false bee62f3
-   Fix support for null in arrays - fixes #371 b04c853

## v3.2.3 - 23 May 2022

-   Fix Only use setKeepAlive in Deno if available 28fbbaf
-   Fix wrong helper match on multiple occurances 02f3854

#### Typescript related

-   Fix Deno assertRejects compatibility (#365) 0f0af92
-   Fix include missing boolean type in JSONValue union (#373) 1817387

## v3.2.2 - 15 May 2022

-   Properly handle errors thrown on commit 99ddae4

## v3.2.1 - 15 May 2022

-   Exclude target_session_attrs from connection obj 43f1442

## v3.2.0 - 15 May 2022

-   Add `sslmode=verify-full` support e67da29
-   Add support for array of fragments 342bf55
-   Add uri decode of host in url - fixes #346 1adc113
-   Add passing of rest url params to connection (ootb support cockroach urls) 41ed84f
-   Fix Deno partial writes 452a30d
-   Fix `as` dynamic helper 3300c40
-   Fix some nested fragments usage 9bfa902
-   Fix missing columns on `Result` when using simple protocol - fixes #350 1e2e298
-   Fix fragments in transactions - fixes #333 75914c7

#### Typescript related

-   Upgrade/fix types (#357) 1e6d312
-   Add optional `onlisten` callback to `listen()` on TypeScript (#360) 6b749b2
-   Add implicit custom type inference (#361) 28512bf
-   Fix and improve sql() helper types (#338) c1de3d8
-   Fix update query type def for `.writable()` and `.readable()` to return promises (#347) 51269ce
-   Add bigint to typescript Serializable - fixes #330 f1e41c3

## v3.1.0 - 22 Apr 2022

-   Add close method to close but not end connections forever 94fea8f
-   Add .values() method to return rows as arrays of values 56873c2
-   Support transform.undefined - fixes #314 eab71e5
-   Support nested fragments values and dynamics - fixes #326 86445ca
-   Fix deno close sequence f76af24
-   Fix subscribe reconnect and add onsubscribe method - fixes #315 5097345
-   Deno ts fix - fixes #327 50403a1

## v3.0.6 - 19 Apr 2022

-   Properly close connections in Deno cbc6a75
-   Only write end message if socket is open 13950af
-   Improve query cancellation 01c2c68
-   Use monotonically increasing time for timeout - fixes #316 9d7a21d
-   Add support for dynamic columns with `returning` - fixes #317 04644c0
-   Fix type errors in TypeScript deno projects (#313) 822fb21
-   Execute forEach instantly 44e9fbe

## v3.0.5 - 6 Apr 2022

-   Fix transaction execution timing 28bb0b3
-   Add optional onlisten function to listen 1dc2fd2
-   Fix dynamic in helper after insert #305 4d63a59

## v3.0.4 - 5 Apr 2022

-   Ensure drain only dequeues if ready - fixes #303 2e5f017

## v3.0.3 - 4 Apr 2022

-   Run tests with github actions b536d0d
-   Add custom socket option - fixes #284 5413f0c
-   Fix sql function overload type inference (#294) 3c4e90a
-   Update deno std to 0.132 and enable last tests 50762d4
-   Send proper client-encoding - Fixes #288 e5b8554

## v3.0.2 - 31 Mar 2022

-   Fix BigInt handling 36a70df
-   Fix unsubscribing (#300) b6c597f
-   Parse update properly with identity full - Fixes #296 3ed11e7

## v3.0.1 - 30 Mar 2022

-   Improve connection queue handling + fix leak cee1a57
-   Use publications option - fixes #295 b5ceecc
-   Throw proper query error if destroyed e148a0a
-   Transaction rejects with rethrown error - fixes #289 f7c8ae6
-   Only create origin stacktrace for tagged and debug - fixes #290 a782edf
-   Include types and readme in deno release - fixes #287 9068820
-   Disable fetch_types for Subscribe options 72e0cdb
-   Update TypeScript types with v3 changes (#293) db05836

## v3.0.0 - 24 Mar 2022

This is a complete rewrite to better support all the features that I was trying to get into v2. There are a few breaking changes from v2 beta, which some (myself included) was using in production, so I'm skipping a stable v2 release and going straight to v3.

Here are some of the new things available, but check the updated docs.

-   Dynamic query builder based on raw sql

src/index.js
import os from "os";
import fs from "fs";

import {
    mergeUserTypes,
    inferType,
    Parameter,
    Identifier,
    Builder,
    toPascal,
    pascal,
    toCamel,
    camel,
    toKebab,
    kebab,
    fromPascal,
    fromCamel,
    fromKebab,
} from "./types.js";

import Connection from "./connection.js";
import { Query, CLOSE } from "./query.js";
import Queue from "./queue.js";
import { Errors, PostgresError } from "./errors.js";
import Subscribe from "./subscribe.js";
import largeObject from "./large.js";

Object.assign(Postgres, {
    PostgresError,
    toPascal,
    pascal,
    toCamel,
    camel,
    toKebab,
    kebab,
    fromPascal,
    fromCamel,
    fromKebab,
    BigInt: {
        to: 20,
        from: [20],
        parse: (x) => BigInt(x), // eslint-disable-line
        serialize: (x) => x.toString(),
    },
});

export default Postgres;

/*
  host                 : '',            // Postgres ip address[es] or domain name[s]
  port                 : 5432,          // Postgres server port[s]
  path                 : '',            // unix socket path (usually '/tmp')
  database             : '',            // Name of database to connect to
  username             : '',            // Username of database user
  password             : '',            // Password of database user
  ssl                  : false,         // true, prefer, require, tls.connect options
  max                  : 10,            // Max number of connections
  max_lifetime         : null,          // Max lifetime in seconds (more info below)
  idle_timeout         : 0,             // Idle connection timeout in seconds
  connect_timeout      : 30,            // Connect timeout in seconds
  prepare              : true,          // Automatic creation of prepared statements
  types                : [],            // Array of custom types, see more below
  onnotice             : fn,            // Defaults to console.log
  onparameter          : fn,            // (key, value) when server param change
  debug                : fn,            // Is called with (connection, query, params, types)
  socket               : fn,            // fn returning custom socket to use
  transform            : {
    undefined          : undefined,     // Transforms undefined values (eg. to null)
    column             : fn,            // Transforms incoming column names
    value              : fn,            // Transforms incoming row values
    row                : fn             // Transforms entire rows
  },
  connection           : {
    application_name   : 'postgres.js', // Default application_name
    ...                                 // Other connection parameters
  },
  target_session_attrs : null,          // Use 'read-write' with multiple hosts to
                                        // ensure only connecting to primary
  fetch_types          : true,          // Automatically fetches types on connect
                                        // on initial connection.
*/
function Postgres(unParsedURL, _options) {
    const options = parseOptions(unParsedURL, _options),
        //"" no_subscribe" is undocumented
        /*
             ...options,
        no_subscribe: true,
        transform: { column: {}, value: {}, row: {} },
        max: 1,
        fetch_types: false,
        idle_timeout: null,
        max_lifetime: null,
        connection: {
            ...options.connection,
            replication: "database",
        },
        onclose: async function () {
          */
        subscribe = options.no_subscribe || Subscribe(Postgres, { ...options });

    let ending = false;

    const queries = Queue(),
        connecting = Queue(),
        reserved = Queue(),
        closed = Queue(),
        ended = Queue(),
        open = Queue(),
        busy = Queue(),
        full = Queue(),
        queues = { connecting, reserved, closed, ended, open, busy, full };

    const connections = [...Array(options.max)].map(() =>
        Connection(options, queues, { onopen, onend, onclose }),
    );

    const sql = Sql(handler);

    Object.assign(sql, {
        get parameters() {
            return options.parameters;
        },
        largeObject: largeObject.bind(null, sql),
        subscribe,
        CLOSE,
        END: CLOSE,
        PostgresError,
        options,
        reserve,
        listen,
        begin,
        close,
        end,
    });

    return sql;

    function Sql(handler) {
        handler.debug = options.debug;

        Object.entries(options.types).reduce((acc, [name, type]) => {
            acc[name] = (x) => new Parameter(x, type.to);
            return acc;
        }, typed);

        Object.assign(sql, {
            types: typed,
            typed,
            unsafe,
            notify,
            array,
            json,
            file,
        });

        return sql;

        function typed(value, type) {
            return new Parameter(value, type);
        }

        function sql(strings, ...args) {
            const query =
                strings && Array.isArray(strings.raw)
                    ? new Query(strings, args, handler, cancel)
                    : typeof strings === "string" && !args.length
                    ? new Identifier(
                          options.transform.column.to
                              ? options.transform.column.to(strings)
                              : strings,
                      )
                    : new Builder(strings, args);
            return query;
        }

        function unsafe(string, args = [], options = {}) {
            arguments.length === 2 &&
                !Array.isArray(args) &&
                ((options = args), (args = []));
            const query = new Query([string], args, handler, cancel, {
                prepare: false,
                ...options,
                simple:
                    "simple" in options ? options.simple : args.length === 0,
            });
            return query;
        }

        function file(path, args = [], options = {}) {
            arguments.length === 2 &&
                !Array.isArray(args) &&
                ((options = args), (args = []));
            const query = new Query(
                [],
                args,
                (query) => {
                    fs.readFile(path, "utf8", (err, string) => {
                        if (err) return query.reject(err);

                        query.strings = [string];
                        handler(query);
                    });
                },
                cancel,
                {
                    ...options,
                    simple:
                        "simple" in options
                            ? options.simple
                            : args.length === 0,
                },
            );
            return query;
        }
    }

    async function listen(name, fn, onlisten) {
        const listener = { fn, onlisten };

        const sql =
            listen.sql ||
            (listen.sql = Postgres({
                ...options,
                max: 1,
                idle_timeout: null,
                max_lifetime: null,
                fetch_types: false,
                onclose() {
                    Object.entries(listen.channels).forEach(
                        ([name, { listeners }]) => {
                            delete listen.channels[name];
                            Promise.all(
                                listeners.map((l) =>
                                    listen(name, l.fn, l.onlisten).catch(() => {
                                        /* noop */
                                    }),
                                ),
                            );
                        },
                    );
                },
                onnotify(c, x) {
                    c in listen.channels &&
                        listen.channels[c].listeners.forEach((l) => l.fn(x));
                },
            }));

        const channels = listen.channels || (listen.channels = {}),
            exists = name in channels;

        if (exists) {
            channels[name].listeners.push(listener);
            const result = await channels[name].result;
            listener.onlisten && listener.onlisten();
            return { state: result.state, unlisten };
        }

        channels[name] = {
            result: sql`listen ${sql.unsafe(
                '"' + name.replace(/"/g, '""') + '"',
            )}`,
            listeners: [listener],
        };
        const result = await channels[name].result;
        listener.onlisten && listener.onlisten();
        return { state: result.state, unlisten };

        async function unlisten() {
            if (name in channels === false) return;

            channels[name].listeners = channels[name].listeners.filter(
                (x) => x !== listener,
            );
            if (channels[name].listeners.length) return;

            delete channels[name];
            return sql`unlisten ${sql.unsafe(
                '"' + name.replace(/"/g, '""') + '"',
            )}`;
        }
    }

    async function notify(channel, payload) {
        return await sql`select pg_notify(${channel}, ${"" + payload})`;
    }

    async function reserve() {
        const q = Queue();
        const c = open.length
            ? open.shift()
            : await new Promise((r) => {
                  queries.push({ reserve: r });
                  closed.length && connect(closed.shift());
              });

        move(c, reserved);
        c.reserved = () =>
            q.length ? c.execute(q.shift()) : move(c, reserved);
        c.reserved.release = true;

        const sql = Sql(handler);
        sql.release = () => {
            c.reserved = null;
            onopen(c);
        };

        return sql;

        function handler(q) {
            c.queue === full ? q.push(q) : c.execute(q) || move(c, full);
        }
    }

    async function begin(options, fn) {
        !fn && ((fn = options), (options = ""));
        const queries = Queue();
        let savepoints = 0,
            connection,
            prepare = null;

        try {
            await sql
                .unsafe("begin " + options.replace(/[^a-z ]/gi, ""), [], {
                    onexecute,
                })
                .execute();
            return await scope(connection, fn);
        } catch (error) {
            throw error;
        }

        async function scope(c, fn, name) {
            const sql = Sql(handler);
            sql.savepoint = savepoint;
            sql.prepare = (x) => (prepare = x.replace(/[^a-z0-9$-_. ]/gi));
            let uncaughtError, result;

            name && (await sql`savepoint ${sql(name)}`);
            try {
                result = await new Promise((resolve, reject) => {
                    const x = fn(sql);
                    Promise.resolve(Array.isArray(x) ? Promise.all(x) : x).then(
                        resolve,
                        reject,
                    );
                });

                if (uncaughtError) throw uncaughtError;
            } catch (e) {
                await (name ? sql`rollback to ${sql(name)}` : sql`rollback`);
                throw (
                    (e instanceof PostgresError &&
                        e.code === "25P02" &&
                        uncaughtError) ||
                    e
                );
            }

            if (!name) {
                prepare
                    ? await sql`prepare transaction '${sql.unsafe(prepare)}'`
                    : await sql`commit`;
            }

            return result;

            function savepoint(name, fn) {
                if (name && Array.isArray(name.raw))
                    return savepoint((sql) => sql.apply(sql, arguments));

                arguments.length === 1 && ((fn = name), (name = null));
                return scope(
                    c,
                    fn,
                    "s" + savepoints++ + (name ? "_" + name : ""),
                );
            }

            function handler(q) {
                q.catch((e) => uncaughtError || (uncaughtError = e));
                c.queue === full
                    ? queries.push(q)
                    : c.execute(q) || move(c, full);
            }
        }

        function onexecute(c) {
            connection = c;
            move(c, reserved);
            c.reserved = () =>
                queries.length ? c.execute(queries.shift()) : move(c, reserved);
        }
    }

    function move(c, queue) {
        c.queue.remove(c);
        queue.push(c);
        c.queue = queue;
        queue === open ? c.idleTimer.start() : c.idleTimer.cancel();
        return c;
    }

    function json(x) {
        return new Parameter(x, 3802);
    }

    function array(x, type) {
        if (!Array.isArray(x)) return array(Array.from(arguments));

        return new Parameter(
            x,
            type || (x.length ? inferType(x) || 25 : 0),
            options.shared.typeArrayMap,
        );
    }

    function handler(query) {
        if (ending)
            return query.reject(
                Errors.connection("CONNECTION_ENDED", options, options),
            );

        if (open.length) return go(open.shift(), query);

        if (closed.length) return connect(closed.shift(), query);

        busy.length ? go(busy.shift(), query) : queries.push(query);
    }

    function go(c, query) {
        return c.execute(query) ? move(c, busy) : move(c, full);
    }

    function cancel(query) {
        return new Promise((resolve, reject) => {
            query.state
                ? query.active
                    ? Connection(options).cancel(query.state, resolve, reject)
                    : (query.cancelled = { resolve, reject })
                : (queries.remove(query),
                  (query.cancelled = true),
                  query.reject(
                      Errors.generic(
                          "57014",
                          "canceling statement due to user request",
                      ),
                  ),
                  resolve());
        });
    }

    async function end({ timeout = null } = {}) {
        if (ending) return ending;

        await 1;
        let timer;
        return (ending = Promise.race([
            new Promise(
                (r) =>
                    timeout !== null &&
                    (timer = setTimeout(destroy, timeout * 1000, r)),
            ),
            Promise.all(
                connections
                    .map((c) => c.end())
                    .concat(
                        listen.sql ? listen.sql.end({ timeout: 0 }) : [],
                        subscribe.sql ? subscribe.sql.end({ timeout: 0 }) : [],
                    ),
            ),
        ]).then(() => clearTimeout(timer)));
    }

    async function close() {
        await Promise.all(connections.map((c) => c.end()));
    }

    async function destroy(resolve) {
        await Promise.all(connections.map((c) => c.terminate()));
        while (queries.length)
            queries
                .shift()
                .reject(Errors.connection("CONNECTION_DESTROYED", options));
        resolve();
    }

    function connect(c, query) {
        move(c, connecting);
        c.connect(query);
        return c;
    }

    function onend(c) {
        move(c, ended);
    }

    function onopen(c) {
        if (queries.length === 0) return move(c, open);

        let max = Math.ceil(queries.length / (connecting.length + 1)),
            ready = true;

        while (ready && queries.length && max-- > 0) {
            const query = queries.shift();
            if (query.reserve) return query.reserve(c);

            ready = c.execute(query);
        }

        ready ? move(c, busy) : move(c, full);
    }

    function onclose(c) {
        move(c, closed);
        c.reserved = null;
        options.onclose && options.onclose(c.id);
        queries.length && connect(c, queries.shift());
    }
}

/*
  host                 : '',            // Postgres ip address[es] or domain name[s]
  port                 : 5432,          // Postgres server port[s]
  path                 : '',            // unix socket path (usually '/tmp')
  database             : '',            // Name of database to connect to
  username             : '',            // Username of database user
  password             : '',            // Password of database user
  ssl                  : false,         // true, prefer, require, tls.connect options
  max                  : 10,            // Max number of connections
  max_lifetime         : null,          // Max lifetime in seconds (more info below)
  idle_timeout         : 0,             // Idle connection timeout in seconds
  connect_timeout      : 30,            // Connect timeout in seconds
  prepare              : true,          // Automatic creation of prepared statements
  types                : [],            // Array of custom types, see more below
  onnotice             : fn,            // Defaults to console.log
  onparameter          : fn,            // (key, value) when server param change
  debug                : fn,            // Is called with (connection, query, params, types)
  socket               : fn,            // fn returning custom socket to use
  transform            : {
    undefined          : undefined,     // Transforms undefined values (eg. to null)
    column             : fn,            // Transforms incoming column names
    value              : fn,            // Transforms incoming row values
    row                : fn             // Transforms entire rows
  },
  connection           : {
    application_name   : 'postgres.js', // Default application_name
    ...                                 // Other connection parameters
  },
  target_session_attrs : null,          // Use 'read-write' with multiple hosts to
                                        // ensure only connecting to primary
  fetch_types          : true,          // Automatically fetches types on connect
                                        // on initial connection.
*/
// a = raw URL string
// b = options
function parseOptions(a, b) {
    if (a && a.shared) return a;

    const env = process.env, // eslint-disable-line
        o = (typeof a === "string" ? b : a) || {},
        { url, multihost } = parseUrl(a),
        query = [...url.searchParams].reduce(
            (a, [b, c]) => ((a[b] = c), a),
            {},
        ),
        host =
            o.hostname ||
            o.host ||
            multihost ||
            url.hostname ||
            env.PGHOST ||
            "localhost",
        port = o.port || url.port || env.PGPORT || 5432,
        user =
            o.user ||
            o.username ||
            url.username ||
            env.PGUSERNAME ||
            env.PGUSER ||
            osUsername();

    o.no_prepare && (o.prepare = false);
    query.sslmode && ((query.ssl = query.sslmode), delete query.sslmode);
    "timeout" in o &&
        (console.log(
            "The timeout option is deprecated, use idle_timeout instead",
        ),
        (o.idle_timeout = o.timeout)); // eslint-disable-line

    const ints = [
        "idle_timeout",
        "connect_timeout",
        "max_lifetime",
        "max_pipeline",
        "backoff",
        "keep_alive",
    ];
    const defaults = {
        max: 10,
        ssl: false,
        idle_timeout: null,
        connect_timeout: 30,
        max_lifetime: max_lifetime,
        max_pipeline: 100,
        backoff: backoff,
        keep_alive: 60,
        prepare: true,
        debug: false,
        fetch_types: true,
        publications: "alltables",
        target_session_attrs: null,
    };

    return {
        host: Array.isArray(host)
            ? host
            : host.split(",").map((x) => x.split(":")[0]),
        port: Array.isArray(port)
            ? port
            : host.split(",").map((x) => parseInt(x.split(":")[1] || port)),
        path: o.path || (host.indexOf("/") > -1 && host + "/.s.PGSQL." + port),
        database:
            o.database ||
            o.db ||
            (url.pathname || "").slice(1) ||
            env.PGDATABASE ||
            user,
        user: user,
        pass: o.pass || o.password || url.password || env.PGPASSWORD || "",
        ...Object.entries(defaults).reduce((acc, [k, d]) => {
            const value =
                k in o
                    ? o[k]
                    : k in query
                    ? query[k] === "disable" || query[k] === "false"
                        ? false
                        : query[k]
                    : env["PG" + k.toUpperCase()] || d;
            acc[k] =
                typeof value === "string" && ints.includes(k) ? +value : value;
            return acc;
        }, {}),
        connection: {
            application_name: "postgres.js",
            ...o.connection,
            ...Object.entries(query).reduce(
                (acc, [k, v]) => (k in defaults || (acc[k] = v), acc),
                {},
            ),
        },
        types: o.types || {},
        target_session_attrs: tsa(o, url, env),
        onnotice: o.onnotice,
        onnotify: o.onnotify,
        onclose: o.onclose,
        onparameter: o.onparameter,
        socket: o.socket,
        transform: parseTransform(o.transform || { undefined: undefined }),
        parameters: {},
        shared: { retries: 0, typeArrayMap: {} },
        ...mergeUserTypes(o.types),
    };
}

function tsa(o, url, env) {
    const x =
        o.target_session_attrs ||
        url.searchParams.get("target_session_attrs") ||
        env.PGTARGETSESSIONATTRS;
    if (
        !x ||
        [
            "read-write",
            "read-only",
            "primary",
            "standby",
            "prefer-standby",
        ].includes(x)
    )
        return x;

    throw new Error("target_session_attrs " + x + " is not supported");
}

function backoff(retries) {
    return (0.5 + Math.random() / 2) * Math.min(3 ** retries / 100, 20);
}

function max_lifetime() {
    return 60 * (30 + Math.random() * 30);
}

function parseTransform(x) {
    return {
        undefined: x.undefined,
        column: {
            from:
                typeof x.column === "function"
                    ? x.column
                    : x.column && x.column.from,
            to: x.column && x.column.to,
        },
        value: {
            from:
                typeof x.value === "function"
                    ? x.value
                    : x.value && x.value.from,
            to: x.value && x.value.to,
        },
        row: {
            from: typeof x.row === "function" ? x.row : x.row && x.row.from,
            to: x.row && x.row.to,
        },
    };
}

// use URL from V8 engine
// done
function parseUrl(url) {
    if (typeof url !== "string") return { url: { searchParams: new Map() } };

    let host = url;
    host = host.slice(host.indexOf("://") + 3).split(/[?/]/)[0];
    host = decodeURIComponent(host.slice(host.indexOf("@") + 1));

    const urlObj = new URL(url.replace(host, host.split(",")[0]));

    return {
        url: {
            username: decodeURIComponent(urlObj.username),
            password: decodeURIComponent(urlObj.password),
            host: urlObj.host,
            hostname: urlObj.hostname,
            port: urlObj.port,
            pathname: urlObj.pathname,
            searchParams: urlObj.searchParams,
        },
        multihost: host.indexOf(",") > -1 && host,
    };
}
//done
function osUsername() {
    try {
        return os.userInfo().username; // eslint-disable-line
    } catch (_) {
        return process.env.USERNAME || process.env.USER || process.env.LOGNAME; // eslint-disable-line
    }
}
